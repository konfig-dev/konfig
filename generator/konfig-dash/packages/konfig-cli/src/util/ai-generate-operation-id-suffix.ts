import { Operation, OperationObject } from 'konfig-lib'
import OpenAI from 'openai'

export async function aiGenerateOperationIdSuffix({
  prefix,
  path,
  method,
  operation,
  operations,
}: {
  prefix: string
  path: string
  method: string
  operation: Operation
  operations: OperationObject[]
}): Promise<string> {
  const openai = new OpenAI({ apiKey: process.env['OPENAI_API_KEY'] })

  const existingIds = operations.map((o) => o.operation.operationId)
  const existingIdsStr = existingIds.join(', ').slice(0, -2)
  const systemPrompt = `You are a generator of operation IDs for an OpenAPI specification. You generate operation IDs in json format: "{ prefix: ... , suffix: ...}"`
  const prompt = generatePrompt({
    prefix,
    path,
    method,
    operation,
    existingIdsStr,
  })

  const response = (
    await openai.chat.completions.create({
      response_format: { type: 'json_object' },
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: prompt },
      ],
      model: 'gpt-3.5-turbo-1106',
    })
  ).choices[0].message.content

  if (response === undefined || response === null || response === '')
    throw Error('OpenAI returned an empty response')

  const responseJson = JSON.parse(response)
  if (responseJson === undefined || responseJson === null)
    throw Error('OpenAI returned an invalid response: it is not a valid json')
  if (responseJson.suffix === undefined || responseJson.suffix === null)
    throw Error('OpenAI returned an invalid json: it does not contain a suffix')

  // Post-process to ensure that rules are followed
  let suffix = responseJson.suffix.replace('_', '')
  suffix = suffix.charAt(0).toLowerCase() + suffix.slice(1)
  suffix = suffix.replace('retrieve', 'get')
  suffix = suffix.replace('fetch', 'get')
  console.log(`AI-generated operation id: ${prefix}${suffix}`)
  return suffix
}

function generatePrompt({
  prefix,
  path,
  method,
  operation,
  existingIdsStr,
}: {
  prefix: string
  path: string
  method: string
  operation: Operation
  existingIdsStr: string
}): string {
  return `Generate an operation ID for this operation in my OpenAPI specification:
Path: ${path}
Method: ${method}
Summary: ${operation.summary}
Description: ${operation.description}

The operation ID is in the format "prefix_suffix". The prefix is: ${prefix}. Generate the suffix and respond in json format.
The operation ID MUST follow ALL of these rules:

1. The suffix must be at least 3 characters long.
2. The suffix must be in camelCase and CANNOT include any underscores.
3. The suffix must not redundantly re-use words from the prefix. (For example: Store_get is preferred over Store_getStore.)
4. If the suffix contains both a verb and a noun, the noun should generally follow the verb. (For example, Store_placeOrder is better than Store_orderPlace).
5. Prefer simple verbs like "get", "create", "update", "delete", and "list" when possible.

Existing operation ids for this OpenAPI specification are as follows: ${existingIdsStr}.
The operation ID MUST be unique within this list of existing operation IDs. You can also use them as a reference to generate this operation id.`
}
