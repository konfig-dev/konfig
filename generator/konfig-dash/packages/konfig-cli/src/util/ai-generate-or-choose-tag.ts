import { Operation, TagObject, OperationObject } from 'konfig-lib'
import OpenAI from 'openai'

export async function aiGenerateOrChooseTag(
  operation: Operation,
  path: string,
  method: string,
  tags: TagObject[]
): Promise<string> {
  const openai = new OpenAI({ apiKey: process.env['OPENAI_API_KEY'] })
  const tagNames = tags.map((t) => t.name)

  // First, try and choose a tag programmatically without using OpenAI
  const tag = programaticallyChooseTag(operation, path, tagNames)
  if (tag !== undefined) return tag
  const systemPrompt = `You are a generator of tags for an OpenAPI specification. You generate tags in json format: "{ tag: ... }"`
  const prompt = generatePrompt(path, method, operation, tagNames)
  const response = (
    await openai.chat.completions.create({
      response_format: { type: 'json_object' },
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: prompt },
      ],
      model: 'gpt-3.5-turbo-1106',
    })
  ).choices[0].message.content

  if (response === undefined || response === null || response === '')
    throw Error('OpenAI returned an empty response')

  const responseJson = JSON.parse(response)
  if (responseJson === undefined || responseJson === null)
    throw Error('OpenAI returned an invalid response: it is not a valid json')
  if (responseJson.tag === undefined || responseJson.tag === null)
    throw Error('OpenAI returned an invalid json: it does not contain a tag')

  tags.push({ name: responseJson.tag })
  return responseJson.tag
}

function programaticallyChooseTag(
  operation: Operation,
  path: string,
  tagNames: string[]
): string | undefined {
  for (const tag of tagNames) {
    if (path.toLowerCase().includes(tag.toLowerCase())) return tag
    if (
      operation.summary &&
      operation.summary.toLowerCase().includes(tag.toLowerCase())
    )
      return tag
  }
  return undefined
}

function generatePrompt(
  path: string,
  method: string,
  operation: Operation,
  tagNames: string[]
): string {
  const tagNamesStr = tagNames.join(', ').slice(0, -2)
  return `The following operation from my OpenAPI specification is missing a tag:
Path: ${path}
Method: ${method}
Summary: ${operation.summary}

Based on this information, choose the best tag from this list of existing tags on the OpenApi Specification: ${tagNamesStr}.
IF AND ONLY IF none of these existing tags are suitable, then generate a new tag. The new tag MUST be a single word, MUST be a noun, and MUST be singular. It MUST NOT be a combination of multiple words.

In general, a good tag name is related to the type of data being operated on (such as "Student"), or the type of operation being performed (such as "Authentication").

Respond with just the tag in json format: "{ tag: ... }" with no other information.`
}
