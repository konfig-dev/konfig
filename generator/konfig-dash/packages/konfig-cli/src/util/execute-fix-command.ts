import * as fs from 'fs-extra'
import {
  isJsonString,
  OBJECT_WITH_NO_PROPERTIES_RULE_NAME,
  parseSpec,
  filterSpecPaths,
  POTENTIAL_INCORRECT_DATA_TYPE_RULE_NAME,
} from 'konfig-lib'
import boxen from 'boxen'
import { Progress } from './fix-progress'
import { oasYamlDump } from './oas-yaml-dump'
import { fixOas } from './fix-oas'
import path from 'path'
import { oasJsonDump } from './oas-json-dump'
import { parseKonfigYaml } from './parse-konfig-yaml'

interface FixOptions {
  format?: boolean
  specInputPath?: string
  spec?: string
  auto?: boolean
  konfigDir?: string
  skipMissingResponseDescriptionFix?: boolean
  skipListUsageSecurity?: boolean
  alwaysYes?: boolean
  ci?: boolean
}

export async function executeFixCommand(options: FixOptions): Promise<void> {
  const flags = {
    ...options,
    auto: options.auto ?? true,
    alwaysYes: !!(options.alwaysYes == null || options.alwaysYes || options.ci), // ci mode should always confirm with yes
    ci: options.ci ?? false,
  }

  const { parsedKonfigYaml } = parseKonfigYaml({
    configDir: flags.konfigDir ?? process.cwd(),
  })

  // We only want to run fix in CI if the konfig.yaml has a specInputPath
  if (flags.ci && parsedKonfigYaml.specInputPath === undefined) {
    console.log(`No specInputPath found in konfig.yaml, skipping fix`)
    return
  }

  if (flags.spec === undefined) {
    flags.spec = parsedKonfigYaml.specPath
    flags.specInputPath = parsedKonfigYaml.specInputPath ?? flags.spec
  }

  if (flags.spec === undefined)
    throw Error(
      `Either specify path to OAS with -s flag or assign "specPath" field in "konfig.yaml"`
    )

  if (flags.specInputPath === undefined) throw Error(`This shouldn't happen`)

  const prependKonfigDir = ({ specPath }: { specPath: string }) => {
    return flags.konfigDir !== undefined
      ? path.join(flags.konfigDir, specPath)
      : specPath
  }

  const specInputPath = prependKonfigDir({ specPath: flags.specInputPath })
  const specOutputPath = prependKonfigDir({ specPath: flags.spec })
  const rawSpec = fs.readFileSync(specInputPath, 'utf-8')
  let spec = await parseSpec(rawSpec)

  // if konfig yaml filters any paths from the spec, remove them now
  if (parsedKonfigYaml.filterPaths) {
    filterSpecPaths({ spec: spec.spec, filter: parsedKonfigYaml.filterPaths })
  }

  if (flags.format) {
    fs.writeFileSync(flags.spec, oasYamlDump(spec))
    return
  }

  const progress: Progress = Progress.getSaved({
    konfigDir: flags.konfigDir ?? process.cwd(),
    progressYamlPathOverride: parsedKonfigYaml.progressYamlPath,
  })

  const {
    issuesFixed,
    infoDescriptionFixed,
    numberOfMissingResponseDescriptionsAdded,
    numberOfExamplesFixed,
    numberOfUpdatedOperationIds,
    numberOfMissingTags,
    numberOfTrailingSlashesFixed,
    numberOfMissing200ResponsesAdded,
    numberOfParametersConvertedToSecurityRequirements,
    numberOfEmptyRequestBodiesRemoved,
    numberOfListUsagesOfSecurityFixed,
    numberOfImproperlyNamedTags,
    numberOfRedundantSecurityAndParametersFixed,
    numberOfObjectsWithNoPropertiesFixed,
    numberOfSchemasDefined,
    numberOfParametersRemovedForNewSecurityScheme,
    numberOfUnstructuredRequestBodiesFixed,
    numberOfIgnorePotentialIncorrectTypeAdded,
    numberOfCustomModifications,
    numberOfIgnoreObjectsWithNoPropertiesAdded,
    numberOfEmptyResponseBodySchemasFixed,
    numberOfDuplicateTagNamesFixed,
    numberOfInvalidServerUrlsFixed,
    numberOfRedundantSecuritySchemesRemoved,
    numberOfExampleAndExamplesFixed,
    numberOfDisallowedHeaderNamesRemoved,
    numberOfNewTagNames,
    numberOfAnyOfTypeNullUsagesReverted,
    numberOfOas31UsagesFixed,
    numberOfExamplesUsageRemoved,
  } = await fixOas({
    spec,
    progress,
    konfigYaml: parsedKonfigYaml,
    alwaysYes: flags.alwaysYes,
    auto: flags.auto,
    ci: flags.ci,
    skipMissingResponseDescription: flags.skipMissingResponseDescriptionFix,
    skipFixListUsageSecurity: flags.skipListUsageSecurity,
  })

  fs.writeFileSync(
    specOutputPath,
    isJsonString(rawSpec) ? oasJsonDump(spec.spec) : oasYamlDump(spec.spec)
  )

  console.log(
    boxen(
      `Updated ${numberOfUpdatedOperationIds} operation IDs
Made ${numberOfCustomModifications} custom modifications
Renamed ${numberOfImproperlyNamedTags} improperly named tags
Removed ${numberOfDisallowedHeaderNamesRemoved} disallowed headers removed
Added ${numberOfIgnoreObjectsWithNoPropertiesAdded} ${OBJECT_WITH_NO_PROPERTIES_RULE_NAME} ignore rule
Added ${numberOfIgnorePotentialIncorrectTypeAdded} ${POTENTIAL_INCORRECT_DATA_TYPE_RULE_NAME} ignore rule
Added ${infoDescriptionFixed} missing info description
Added ${numberOfMissingResponseDescriptionsAdded} missing response descriptions
Added ${numberOfMissingTags} missing tags
Added ${numberOfEmptyResponseBodySchemasFixed} empty response body schemas
Added ${numberOfSchemasDefined} named schemas
Added ${numberOfMissing200ResponsesAdded} missing 2xx responses
Added ${numberOfNewTagNames} new tag names
Fixed ${numberOfDuplicateTagNamesFixed} duplicate tag names fixed
Fixed ${numberOfInvalidServerUrlsFixed} invalid server URLs fixed
Fixed ${numberOfUnstructuredRequestBodiesFixed} unstructured request bodies
Fixed ${numberOfObjectsWithNoPropertiesFixed} objects with no properties
Fixed ${numberOfExampleAndExamplesFixed} redundant "example" & "examples" fields
Fixed ${numberOfExamplesFixed} examples with invalid schemas
Fixed ${numberOfParametersConvertedToSecurityRequirements} parameters that should be security requirements
Fixed ${numberOfListUsagesOfSecurityFixed} list usages of security
Removed ${numberOfTrailingSlashesFixed} trailing slashes
Removed ${numberOfRedundantSecurityAndParametersFixed} redundant security requirement and parameters
Removed ${numberOfRedundantSecuritySchemesRemoved} redundant security schemes
Removed ${numberOfParametersRemovedForNewSecurityScheme} parameters replace with security requirement
Removed ${numberOfEmptyRequestBodiesRemoved} empty request bodies
Fixed ${numberOfAnyOfTypeNullUsagesReverted} anyOf type null usages
Fixed ${numberOfOas31UsagesFixed} OAS 3.1.x usages
Fixed ${numberOfExamplesUsageRemoved} "examples" usage`,
      {
        title: `Fixed ${issuesFixed} Issues`,
        borderColor: issuesFixed === 0 ? 'gray' : 'green',
        padding: 1,
      }
    )
  )
}
