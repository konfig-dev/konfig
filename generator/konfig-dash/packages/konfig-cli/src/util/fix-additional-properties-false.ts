import { Spec, recurseObject } from 'konfig-lib'

/**
 * "additionalProperties: false" causes issue when mocking an API.  The
 * following type of errors will appear for no reason: "âœ–  error     Request
 * body must NOT have additional properties; found 'userId'" but userId is
 * actually a property in the schema so the error is just wrong.  To get around
 * this I added this rule to remove "additionalProperties: false" from the
 * schema.
 */
export function fixAdditionalPropertiesFalse({ spec }: { spec: Spec }): number {
  let numberOfAdditionalPropertiesFixed = 0
  recurseObject(spec.spec, ({ value: schema }) => {
    if (typeof schema !== 'object') return
    if (schema === null) return
    if (schema.type !== 'object') return
    if (schema.additionalProperties === undefined) return
    if (schema.additionalProperties === true) return
    delete schema.additionalProperties
    numberOfAdditionalPropertiesFixed++
  })
  return numberOfAdditionalPropertiesFixed
}
