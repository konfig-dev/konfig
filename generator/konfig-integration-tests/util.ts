import { execa } from "execa";
import * as path from "path";
import yaml from "js-yaml";
import { expect } from "vitest";
import * as fs from "fs";
import {
  KONFIG_YAML_NAME,
  KonfigYaml,
  KonfigYamlAdditionalGeneratorConfig,
  KonfigYamlGeneratorConfig,
  KonfigYamlType,
} from "konfig-lib";
import waitOn from "wait-on";

const USE_RUN =
  process.env.GITHUB_ACTIONS === "true" ||
  (process.env.RUN && process.env.RUN.trim() !== "");

console.log("USE_RUN", USE_RUN);

// relative path from this file to: ""../konfig-dash/packages/konfig-cli/bin/dev""
// use nodejs __dirname to get the current directory of this file
const KONFIG_CLI_PATH = path.join(
  __dirname,
  `../konfig-dash/packages/konfig-cli/bin/${USE_RUN ? "run" : "dev"}`
);

/**
 * parse 'python-pydantic-empty-response' from 'tests/python-pydantic-empty-response.test.ts > python-pydantic-empty-response'
 */
function getCurrentTestName(): string {
  const currentTestName = expect.getState().currentTestName;
  const parsed = currentTestName?.split(" > ")[1];
  if (!parsed) {
    throw new Error(`Unable to parse current test name: ${currentTestName}`);
  }
  return parsed;
}

export async function e2e(
  mockServerPort: number,
  customAssertions?: () => void
) {
  if (!getCurrentTestName()) {
    throw new Error("Unable to get current test name");
  }
  const sdkDir = path.join(__dirname, "sdks", getCurrentTestName());

  const apiUrl = process.env.KONFIG_API_URL ?? "http://127.0.0.1:8911";

  // wait until KONFIG_API_URL/healthz is available
  await waitOn({
    resources: [`${apiUrl}/healthz`],
    timeout: 100000,
  });

  const env = {
    // This ensure oclif doesn't bug out
    // See: https://github.com/oclif/oclif/issues/1161#issuecomment-1661372245
    NODE_ENV: "",
  };

  // run "konfig generate -d -I" inside the path
  await execa(KONFIG_CLI_PATH, ["generate", "-d", "-I"], {
    cwd: sdkDir,
    stdio: "inherit",
    env,
  });

  // run "konfig test" inside the path
  await execa(KONFIG_CLI_PATH, ["test", "-p", mockServerPort.toString()], {
    cwd: sdkDir,
    stdio: "inherit",
    env,
  });

  // validate top-level README.md
  const konfigYamlPath = path.join(sdkDir, KONFIG_YAML_NAME);
  if (!fs.existsSync(konfigYamlPath))
    throw new Error(
      `Make sure you are in same directory as '${KONFIG_YAML_NAME}'`
    );
  const konfigYamlRaw = fs.readFileSync(konfigYamlPath, "utf-8");
  const loadedKonfigYaml: KonfigYamlType = yaml.load(
    konfigYamlRaw
  ) as KonfigYamlType;
  const parsedKonfigYaml = KonfigYaml.parse(loadedKonfigYaml);

  const generators: (
    | KonfigYamlGeneratorConfig
    | KonfigYamlAdditionalGeneratorConfig
  )[] = [
    ...Object.values(parsedKonfigYaml.generators),
    ...(parsedKonfigYaml.additionalGenerators !== undefined
      ? Object.values(parsedKonfigYaml.additionalGenerators)
      : []),
  ];
  for (const generator of generators) {
    expect(
      normalizeDocumentation(
        fs.readFileSync(
          path.join(sdkDir, generator.outputDirectory, "README.md"),
          "utf-8"
        )
      )
    ).toMatchSnapshot();
  }

  if (customAssertions) {
    customAssertions();
  }
}

// Removes the [GitHub last commit] line from the README
function normalizeDocumentation(readme: string) {
  // matches [GitHub last commit] to the end of the line
  const pattern = /\[!\[GitHub last commit\].*$\n?/gm;
  return readme.replace(pattern, "");
}
