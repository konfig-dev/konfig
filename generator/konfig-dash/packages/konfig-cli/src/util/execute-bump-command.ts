import { CliUx } from '@oclif/core'
import { getPackageVersion, setPackageVersion, KonfigYamlType } from 'konfig-lib'
import { generateKonfigYamlString } from '../util/generate-konfig-yaml-string'
import { incrementVersionWithPreservedTag } from '../util/increment-version-with-preserved-tag'
import semver from 'semver'
import * as fs from 'fs-extra'

export async function executeBumpCommand({
  filter,
  releaseType,
  loadedKonfigYaml,
  konfigYamlPath,
}: {
  filter: string[] | null
  releaseType: semver.ReleaseType
  loadedKonfigYaml: KonfigYamlType
  konfigYamlPath: string
}): Promise<void> {

  if (filter !== null)
    CliUx.ux.info(`Performing ${releaseType} bump of ${filter.join(', ')}`)
  else CliUx.ux.info(`Performing "${releaseType}" bump`)

  for (const [generatorName, generatorConfig] of [
    ...Object.entries(loadedKonfigYaml.generators),
    ...Object.entries(loadedKonfigYaml.additionalGenerators ?? []),
  ]) {
    if (filter !== null && !filter.includes(generatorName)) continue
    if ('disabled' in generatorConfig && generatorConfig.disabled) continue
    const currentVersion = getPackageVersion({ generatorConfig })
    const newVersion = incrementVersionWithPreservedTag(currentVersion, releaseType)
    CliUx.ux.info(`Bumping ${generatorName} SDK from ${currentVersion} --> ${newVersion}`)
    if (newVersion === null)
      CliUx.ux.error(`Could not compute new version for ${generatorName} SDK currently at ${currentVersion}`)
    setPackageVersion({ generatorConfig, version: newVersion })
  }
  CliUx.ux.action.start('Writing new version to konfig.yaml')
  fs.writeFileSync(konfigYamlPath, generateKonfigYamlString(loadedKonfigYaml))
  CliUx.ux.action.stop()
}
