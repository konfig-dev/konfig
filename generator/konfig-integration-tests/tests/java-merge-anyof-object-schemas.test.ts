import { e2e } from "../util";
import { test } from "vitest";

test("java-merge-anyof-object-schemas", async () => {
  await e2e(4034);
});