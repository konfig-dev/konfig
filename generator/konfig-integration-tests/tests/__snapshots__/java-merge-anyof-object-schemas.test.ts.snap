// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`java-merge-anyof-object-schemas 1`] = `
"# java-merge-anyof-object-schemas<a id="java-merge-anyof-object-schemas"></a>

A simple API based for testing java-merge-anyof-object-schemas.

[![Maven Central](https://img.shields.io/badge/Maven%20Central-v1.0.0-blue)](https://central.sonatype.com/artifact/com.konfigthis/java-merge-anyof-object-schemas/1.0.0)
[![More Info](https://img.shields.io/badge/More%20Info-Click%20Here-orange)](http://example.com/support)

## Table of Contents<a id="table-of-contents"></a>

<!-- toc -->

- [Requirements](#requirements)
- [Installation](#installation)
  * [Maven users](#maven-users)
  * [Gradle users](#gradle-users)
  * [Android users](#android-users)
  * [Others](#others)
- [Getting Started](#getting-started)
- [Reference](#reference)
  * [\`javamergeanyofobjectschemasclient.test.fetch\`](#javamergeanyofobjectschemasclienttestfetch)
  * [\`javamergeanyofobjectschemasclient.test.mergeDifferent\`](#javamergeanyofobjectschemasclienttestmergedifferent)
  * [\`javamergeanyofobjectschemasclient.test.mergeSame\`](#javamergeanyofobjectschemasclienttestmergesame)

<!-- tocstop -->

## Requirements<a id="requirements"></a>

Building the API client library requires:

1. Java 1.8+
2. Maven (3.8.3+)/Gradle (7.2+)

If you are adding this library to an Android Application or Library:

3. Android 8.0+ (API Level 26+)

## Installation<a id="installation"></a>

To install the API client library to your local Maven repository, simply execute:

\`\`\`shell
mvn clean install
\`\`\`

To deploy it to a remote Maven repository instead, configure the settings of the repository and execute:

\`\`\`shell
mvn clean deploy
\`\`\`

Refer to the [OSSRH Guide](http://central.sonatype.org/pages/ossrh-guide.html) for more information.

### Maven users<a id="maven-users"></a>

Add this dependency to your project's POM:

\`\`\`xml
<dependency>
  <groupId>com.konfigthis</groupId>
  <artifactId>java-merge-anyof-object-schemas</artifactId>
  <version>1.0.0</version>
  <scope>compile</scope>
</dependency>
\`\`\`

### Gradle users<a id="gradle-users"></a>

Add this dependency to your \`build.gradle\`:

\`\`\`groovy
// build.gradle
repositories {
  mavenCentral()
}

dependencies {
   implementation "com.konfigthis:java-merge-anyof-object-schemas:1.0.0"
}
\`\`\`

### Android users<a id="android-users"></a>

Make sure your \`build.gradle\` file as a \`minSdk\` version of at least 26:
\`\`\`groovy
// build.gradle
android {
    defaultConfig {
        minSdk 26
    }
}
\`\`\`

Also make sure your library or application has internet permissions in your \`AndroidManifest.xml\`:

\`\`\`xml
<!--AndroidManifest.xml-->
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:tools="http://schemas.android.com/tools">
    <uses-permission android:name="android.permission.INTERNET"/>
</manifest>
\`\`\`

### Others<a id="others"></a>

At first generate the JAR by executing:

\`\`\`shell
mvn clean package
\`\`\`

Then manually install the following JARs:

* \`target/java-merge-anyof-object-schemas-1.0.0.jar\`
* \`target/lib/*.jar\`

## Getting Started<a id="getting-started"></a>

Please follow the [installation](#installation) instruction and execute the following Java code:

\`\`\`java
import com.konfigthis.javamergeanyofobjectschemas.client.ApiClient;
import com.konfigthis.javamergeanyofobjectschemas.client.ApiException;
import com.konfigthis.javamergeanyofobjectschemas.client.ApiResponse;
import com.konfigthis.javamergeanyofobjectschemas.client.JavaMergeAnyofObjectSchemasClient;
import com.konfigthis.javamergeanyofobjectschemas.client.Configuration;
import com.konfigthis.javamergeanyofobjectschemas.client.auth.*;
import com.konfigthis.javamergeanyofobjectschemas.client.model.*;
import com.konfigthis.javamergeanyofobjectschemas.client.api.TestApi;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public class Example {
  public static void main(String[] args) {
    Configuration configuration = new Configuration();
    configuration.host = "https://java-merge-anyof-object-schemas.konfigthis.com";
    
    configuration.apiKey  = "YOUR API KEY";
    JavaMergeAnyofObjectSchemasClient client = new JavaMergeAnyofObjectSchemasClient(configuration);
    try {
      ABC result = client
              .test
              .fetch()
              .execute();
      System.out.println(result);
      System.out.println(result.getA());
      System.out.println(result.getB());
      System.out.println(result.getC());
    } catch (ApiException e) {
      System.err.println("Exception when calling TestApi#fetch");
      System.err.println("Status code: " + e.getStatusCode());
      System.err.println("Reason: " + e.getResponseBody());
      System.err.println("Response headers: " + e.getResponseHeaders());
      e.printStackTrace();
    }

    // Use .executeWithHttpInfo() to retrieve HTTP Status Code, Headers and Request
    try {
      ApiResponse<ABC> response = client
              .test
              .fetch()
              .executeWithHttpInfo();
      System.out.println(response.getResponseBody());
      System.out.println(response.getResponseHeaders());
      System.out.println(response.getStatusCode());
      System.out.println(response.getRoundTripTime());
      System.out.println(response.getRequest());
    } catch (ApiException e) {
      System.err.println("Exception when calling TestApi#fetch");
      System.err.println("Status code: " + e.getStatusCode());
      System.err.println("Reason: " + e.getResponseBody());
      System.err.println("Response headers: " + e.getResponseHeaders());
      e.printStackTrace();
    }
  }
}

\`\`\`
## Reference<a id="reference"></a>


### \`javamergeanyofobjectschemasclient.test.fetch\`<a id="javamergeanyofobjectschemasclienttestfetch"></a>

Provide an input parameter to receive a JSON value with properties.

#### üõ†Ô∏è Usage<a id="üõ†Ô∏è-usage"></a>

\`\`\`java
ABC result = client
        .test
        .fetch()
        .execute();
\`\`\`

#### üîÑ Return<a id="üîÑ-return"></a>

[ABC](./src/main/java/com/konfigthis/client/model/ABC.java)

#### üåê Endpoint<a id="üåê-endpoint"></a>

\`/simple-endpoint\` \`GET\`

[üîô **Back to Table of Contents**](#table-of-contents)

---


### \`javamergeanyofobjectschemasclient.test.mergeDifferent\`<a id="javamergeanyofobjectschemasclienttestmergedifferent"></a>

merge different

#### üõ†Ô∏è Usage<a id="üõ†Ô∏è-usage"></a>

\`\`\`java
A1 result = client
        .test
        .mergeDifferent()
        .execute();
\`\`\`

#### üîÑ Return<a id="üîÑ-return"></a>

[A1](./src/main/java/com/konfigthis/client/model/A1.java)

#### üåê Endpoint<a id="üåê-endpoint"></a>

\`/merge-different\` \`GET\`

[üîô **Back to Table of Contents**](#table-of-contents)

---


### \`javamergeanyofobjectschemasclient.test.mergeSame\`<a id="javamergeanyofobjectschemasclienttestmergesame"></a>

merge same

#### üõ†Ô∏è Usage<a id="üõ†Ô∏è-usage"></a>

\`\`\`java
B1 result = client
        .test
        .mergeSame()
        .execute();
\`\`\`

#### üîÑ Return<a id="üîÑ-return"></a>

[B1](./src/main/java/com/konfigthis/client/model/B1.java)

#### üåê Endpoint<a id="üåê-endpoint"></a>

\`/merge-same\` \`GET\`

[üîô **Back to Table of Contents**](#table-of-contents)

---


## Author<a id="author"></a>
This Java package is automatically generated by [Konfig](https://konfigthis.com)
"
`;
